lang/syn/src/codegen/accounts/bumps.rs

```diff
@@ -0,0 +1,76 @@
+use crate::{
+    codegen::accounts::{generics, ParsedGenerics},
+    *,
+};
+use std::fmt::Display;
+
+use super::constraints;
+
+pub fn generate_bumps_name<T: Display>(anchor_ident: &T) -> Ident {
+    Ident::new(&format!("{}Bumps", anchor_ident), Span::call_site())
+}
+
+pub fn generate(accs: &AccountsStruct) -> proc_macro2::TokenStream {
+    let name = &accs.ident;
+    let bumps_name = generate_bumps_name(name);
+    let ParsedGenerics {
+        combined_generics,
+        trait_generics: _,
+        struct_generics,
+        where_clause,
+    } = generics(accs);
+
+    let bump_fields: Vec<proc_macro2::TokenStream> = accs
+        .fields
+        .iter()
+        .filter_map(|af| {
+            let ident = af.ident();
+
+            match af {
+                AccountField::Field(f) => {
+                    let constraints = constraints::linearize(&f.constraints);
+                    let bump_field = quote!(pub #ident: u8);
+
+                    for c in constraints.iter() {
+                        // Verify this in super::constraints
+                        // The bump is only cached if
+                        // - PDA is marked as init
+                        // - PDA is not init, but marked with bump without a target
+
+                        match c {
+                            Constraint::Seeds(c) => {
+                                if !c.is_init && c.bump.is_none() {
+                                    return Some(bump_field);
+                                }
+                            }
+                            Constraint::Init(c) => {
+                                if c.seeds.is_some() {
+                                    return Some(bump_field);
+                                }
+                            }
+                            _ => (),
+                        }
+                    }
+                    None
+                }
+                AccountField::CompositeField(s) => {
+                    let comp_bumps_struct = generate_bumps_name(&s.symbol);
+                    let bumps = quote!(pub #ident: #comp_bumps_struct);
+
+                    Some(bumps)
+                }
+            }
+        })
+        .collect();
+
+    quote! {
+        #[derive(Default, Debug)]
```

Comment:
This is the one concern I have with this PR. Deriving `Default` for the bumps struct puts the bumps in an incorrect state during some part of the program execution. Ideally bumps should never be in an incorrect state thoughout the execution of the program.

If we compare the existing implementation(`BTreeMap`) with this implementation(`Bumps` struct):

`BTreeMap<String, u8>`:
- Initializing default `bumps` means initializing an empty `BTreeMap`
- The fields are inserted only when the bump is known which ensures the correctness of the bump

`Bumps` struct:
- Initializing default `bumps` means initializing a struct with account names as the properties and `0`s as values because of the `Default` derivation
- The values of the struct are mutated to the correct bump values at some point in the program lifecycle before the user code

This requires an extra attention to make sure the values are mutated at the right time and it could lead to severe security issues if not done correctly meanwhile the existing implementation would simply error in the case of forgetting to insert the bump.

Reply:
We can remove the Default macro and modify the try_accounts func to return a tuple of Accounts and Bump. What do you think of this approach? In this case we can control the bump to be sure to have canonical bump

Reply:
Sounds promising as it looks like that would also remove the necessity of mutating the `Bumps` struct in `try_accounts` implementation and isolate the creation of bumps in one place instead of two. What's your take?

Reply:
I think it could be a good idea. I hope it won't create too many changes

Reply:
Yes, depends how much change is needed to accomplish this. The current implementation seems to be safe so if it requires a lot of changes, it's probably not worth it and we can merge as is.